src/main/java/de/tum/in/www1/artemis/config/migration/MigrationRegistry.java

```diff
@@ -0,0 +1,170 @@
+package de.tum.in.www1.artemis.config.migration;
+
+import static tech.jhipster.config.JHipsterConstants.SPRING_PROFILE_TEST;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.core.env.Profiles;
+import org.springframework.stereotype.Service;
+import org.thymeleaf.util.StringUtils;
+
+import de.tum.in.www1.artemis.domain.MigrationChangelog;
+import de.tum.in.www1.artemis.repository.MigrationChangeRepository;
+
+/**
+ * This service allows registering certain entries containing functionality that gets executed on application startup. The entries must extend {@link MigrationEntry}.
+ * The class contains additional utility functionality that verifies the registry on each execution to prevent a corruption.
+ */
+@Service
+public class MigrationRegistry {
+
+    private final Logger log = LoggerFactory.getLogger(MigrationRegistry.class);
+
+    @Value("${artemis.version}")
+    private String artemisVersion;
+
+    // Using TreeMap to allow sorting. I'm using a map because with a list entries could accidentally be switched
+    private final TreeMap<Integer, Class<? extends MigrationEntry>> migrationEntryMap = new TreeMap<>();
+
+    private TreeMap<Integer, MigrationEntry> instantiatedMigrationEntryMap;
+
+    private final AutowireCapableBeanFactory beanFactory;
+
+    private final MigrationChangeRepository migrationChangeRepository;
+
+    private final MessageDigest messageDigest = MessageDigest.getInstance("MD5");
+
+    public MigrationRegistry(AutowireCapableBeanFactory beanFactory, MigrationChangeRepository migrationChangeRepository) throws NoSuchAlgorithmException {
+        // Here we define the order of the ChangeEntries
+        // migrationEntryMap.put(0, MigrationEntry20211127_120000.class);
```

Comment:
As far as I understand, the constructor will get longer, and longer over the time.
Maybe consider using the autowire functionallity directly:
https://stackoverflow.com/questions/40286047/spring-get-all-beans-of-certain-interface-and-type
You could order them according to your name schema, also filter them according to certain packages, and do not have to register all of them in the constructor.

Reply:
I agree with you in general. However, I have already experienced how good the naming schema works with Liquibase. That's why I thought about the map to have a failsafe and prevent that someone named it differently by accident.

What do you think?

Reply:
The only thing I'm unsure about is whether it's good that this class will get bigger and bigger over the time.
I would suggest to load the migration entries on the fly. There may be other possibilities to ensure consistency. Maybe add a field "previous migration" to ensure that the order of migrations is consistent.

As alternative, you could log/throw some error if a MigrationChange is found and not conform to the naming scheme.

Or are there any other problems with the naming scheme.

Reply:
The issue is not necessarily the format. But if you write a wrong number, you suddenly end up in a completely different position.

A linked list would be possible, would require sorting at startup.

We could however put the reference to the latest one in the registry and link all other together as linked list where they have the reference on the previous one.
The question would be how to detect if the chain breaks. I would get a possible issue with the tests.
In general this might be tough to test.

Also, same thing with Liquibase, we should develop a concept of not supporting all major versions. So once there is a new major version we should compress all database changes up to the previous one (latest 5.0.0 => compress 0.1.0 to < 4.0.0;) This wouldn't let the file bloat up too much.

So I'd actually suggest to keep it like that. As it's only a hardcoded registry, it could be changed at any point in the future if deemed necessary.